### PR TITLE
Speech Recognition v2 - StartListening/StopListening

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Description of Change ###

StartListeningAsync is to start listening for speech. The method returns "immediately", it doesn't wait for the recognition result. I would not name it Background, to not confuse users, that it works like a background service and allows to spy for the user.
so flow for users:
StartListeningAsync
Subscribe for events (SpeechCompleted/PartialSpeech)
do some other stuff
Whenever I want, I can stop listening without caring for the result.

In future we can extend the API with Pause/Resume async

trying to replicate this API: https://learn.microsoft.com/en-us/uwp/api/windows.media.speechrecognition.speechcontinuousrecognitionsession.startasync?view=winrt-22621

Added new APIs : 
- StartListeting
- StopListening
- SpeechToTextState
- SpeechRecognitionUpdated event
- SpeechRecognitionCompleted event

Removed:
Offline Speech Recognition on Windows. Maybe that is my pronunciation, but the output result is much worse than with online recognition.

 ### Linked Issues ###
 

 - Inspired by https://github.com/CommunityToolkit/Maui/discussions/1210

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [ ] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/315 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs:85]: Why not use the cancellationToken provided by the user?

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.shared.cs:93]: Can those EventArgs be moved to another file?
Also, is a pattern on .NET to finish the name with `Args` to be clear that this model is an `EventArgs`, so consider changing it

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/ISpeechToText.shared.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/ISpeechToText.shared.cs:40]: I'm trying to understand this callback, if I'm using async/await pattern to start/stop listening I would spect that one of those methods (StartListeningAsync or StopListeneningAsync) returns the string result instead of having to subscribe to an eventHandler. Is something that prevents those methods from returning the result?

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs:93]: Why not return the `getRecognitionTaskCompletionSource.Task`?

**Review by pictos — CHANGES_REQUESTED** [review]
This looks good in general, just a friendly reminder this will be a breaking change since you're removing a feature on Windows. Also, I could see a lot of places where the `CancellationToken` provided by the user is being ignored, and this isn't ideal.

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs:85]: cancellation token is not used in any of those methods. removed it at all

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/ISpeechToText.shared.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/ISpeechToText.shared.cs:40]: we already have a method ListenAsync, that returns result and intermediate result. So the idea of these methods just start and stop listening. we may later add Pause and resume for example

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs:93]: this method doesn't use getRecognitionTaskCompletionSource. if you call this method, getRecognitionTaskCompletionSource is null.

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks Vlad! 

Looks like we just need to add unit tests for the new public APIs and write the the docs.

Otherwise, the implementation looks good.

I made a couple tweaks:
- Added `event ISpeechToText.StateChanged` 
- Renamed `ISpeechToText.State` -> `ISpeechToText.CurrentState`
- Renamed `ISpeechToText.StartListeningAsync` -> `ISpeechToText.StartListenAsync`
- Renamed `ISpeechToText.StopListeningAsync` -> `ISpeechToText.StopListenAsync`
- Renamed `OnSpeechToTextRecognitionResultCompleted` -> `SpeechToTextRecognitionResultCompletedEventArgs`
- Renamed `OnSpeechToTextRecognitionResultUpdated` -> `SpeechToTextRecognitionResultUpdatedEventArgs`
- Updated XML Comments for new APIs
- Updated Sample App to be more descriptive 

Screenshot_1694303409

**Comment by brminnick** [comment]
Adding the `do not merge` and `pending documentation` labels until we have the Docs PR opened

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs:100]: Probably I'm not following the flow, so how it's null if it's used by `recognitionTask` at line 69? And also looks like the only place that it gets instantiated is on `InternalListenAsync` method.

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs:100]: it is marked as nullable there.
we have 2 methods: IntenalListenAsync and InternalStartListenAsync.
- InternalListenAsync is waiting for the result, so it needs that getRecognitionTaskCompletionSource. (and it initialize it, then calls InternalStartListenAsync, In that case InternalStartListenAsync has defined getRecognitionTaskCompletionSource)
- InternalStartListenAsync is called directly from new method StartListenAsync, so this getRecognitionTaskCompletionSource is not initialized in that case and just ignored

**Comment by VladislavAntonyuk** [comment]
StateChanged event only makes sense on Android, where we manually switch the state, all other platforms have get only property, that returns the real state of speech recognizer.

**Review by brminnick — DISMISSED** [review]
@JoonghyunCho Hey Jay!

Could you incorporate the newly added event, `ISpeechToText.StateChanged` into the Tizen implementation?

We have a new method, `void OnSpeechToTextStateChanged(SpeechToTextState)`, that you can call any time the `SpeechToText` state changes on Tizen.

https://github.com/CommunityToolkit/Maui/blob/7eda8ffc42ce36b259c04472e147092b7e597d00/src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.shared.cs#L83-L86 

I'm not sure if Tizen has it's own event, like Windows' `SpeechRecognizer.StateChanged` event, that you can subscribe to, or if you'll need to call `OnSpeechToTextStateChanged()` manually, like we're doing for iOS, MacCatalyst and Windows.

**Comment by brminnick** [comment]
Thanks @vlad! I think we're super close 💯

Once Jay implements `StateChanged` for Tizen, we should be ready to merge!

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs:17]: This has the same docs as the `SpeechToTextRecognitionResultCompletedEventArgs`, and `SpeechToTextStateChangedEventArgs` can't we merge this as single one? I know, the name is little different, but the arguments and meanings are the same

Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs:17]: I would keep them separate. The same as in Windows native implementation. They may be extended later with more properties (Choices, recognition accuracy etc)

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs:17]: we can have a base type with those properties, I believe all Args will have the same properties or most of them will be the same

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs:81]: Completing this method synchrony could could be related to this [issue](https://github.com/CommunityToolkit/Maui/issues/1415)

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.android.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.android.cs:70]: this should be disposed

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SharedSpeechToTextImplementation.macios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SharedSpeechToTextImplementation.macios.cs:90]: This should be disposed

**Review by pictos — CHANGES_REQUESTED** [review]
Just found a couple of areas that we can improve and a question regarding an opened issue.

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SharedSpeechToTextImplementation.macios.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SharedSpeechToTextImplementation.macios.cs]: Good catch! Fixed ✅

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.android.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.android.cs]: Good catch! Fixed ✅

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs:81]: What do you think about merging this PR and addressing https://github.com/CommunityToolkit/Maui/issues/1415 in a future PR? 

As long as this PR isn't introducing that bug (or making it worse), I'd like to merge this because  this PR is currently blocking us from releasing CommunityToolkit.Maui v6.0.0.

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/Events/SpeechToTextRecognitionResultUpdatedEventArgs.cs:17]: You both make good points. 

I just played around a bit with this in code and I think it's best to keep the current implementation for now.

We can't create `public abstract class SpeechToTextRecognitionEventArgs` and add `public string RecognitionResult { get; }` to it because then we's have a 3rd `EventArgs`, `SpeechToTextStateChangedEventArgs`, that doesn't implement this base class. Same is true for creating `public interface ISpeechToTextRecognitionEventArgs`; we'd always have a 3rd event args that doesn't implement an abstraction which may lead to confusion for devs expecting it.

There's nothing smelly in this code. But if/when we add additional `EventArgs` to `ISpeechToText`, we'd definitely want to rediscuss this.

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.ios.cs:81]: I'm fine with that, just want to point out if completing in a sync way is the cause of that issue. No problem to merge as it's

**Comment by JoonghyunCho** [comment]
Sorry I'm late!!!!! Just added `StateChanged` to Tizen :)

**Comment by JoonghyunCho** [comment]
Not sure why it suddenly cause the vulnerability error with the SkiaSharp. Can you guess anything?

File: src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs]: Any actions here? This is the final open conversation on this PR before we can merge it. 

Given the [security vulnerability that we patched on SkiaSharp today](https://github.com/CommunityToolkit/Maui/issues/1424), I'd like to publish a new release of `CommunityToolkit.Maui` as quickly as possible so that our users are subsequently patched!

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Essentials/SpeechToText/SpeechToTextImplementation.macos.cs]: For me it's good to go

**Comment by VladislavAntonyuk** [comment]
Thank you @brminnick , @JoonghyunCho  and @pictos !
@brminnick Can we merge a PR?

**Comment by brminnick** [comment]
For sure! I was just giving it a final look right now 🙌

**Review by brminnick — APPROVED** [review]

**Pull request merged**
